src/palimpzest/generators/generators.py

```diff
@@ -0,0 +1,234 @@
+        payload = None
+        if self.model_name == Model.GPT_4V.value:
+            payload = {
+                "model": "gpt-4-vision-preview",
+                "messages": [
+                    {
+                        "role": "user",
+                        "content": [
+                            {
+                                "type": "text",
+                                "text": prompt,
+                            },
+                            {
+                                "type": "image_url",
+                                "image_url": {
+                                    "url": f"data:image/jpeg;base64,{base64_image}"
+                                },
+                            }
+                        ]
+                    }
+                ],
+                "max_tokens": 4000
+            }
+
+        elif self.model_name == Model.GEMINI_1V.value:
+            payload = [prompt, Image.open(io.BytesIO(self._decode_image(base64_image)))]
+
+        else:
+            raise ValueError(f"Model must be one of the image models specified in palimpzest.constants.Model")
+
+        return payload
+
+    def _generate_response(self, client: Union[OpenAI, genai.GenerativeModel], payload: Any) -> Tuple[str, str, dict]:
+        answer, finish_reason, usage = None, None, None
+
+        if self.model_name == Model.GPT_4V.value:
+            completion = client.chat.completions.create(**payload)
+            candidate = completion.choices[-1]
+            answer = candidate.message.content
+            finish_reason = candidate.finish_reason
+            usage = completion.usage
+
+        elif self.model_name == Model.GEMINI_1V.value:
+            response = client.generate_content(payload)
+            candidate = response.candidates[-1]
+            answer = candidate.content.parts[0].text
+            finish_reason = candidate.finish_reason
+            usage = {}
+
+        else:
+            raise ValueError(f"Model must be one of the image models specified in palimpzest.constants.Model")
+
+        return answer, finish_reason, usage
+
+    @retry(
+        wait=wait_exponential(multiplier=RETRY_MULTIPLIER, max=RETRY_MAX_SECS),
+        stop=stop_after_attempt(RETRY_MAX_ATTEMPTS),
+        after=log_attempt_number,
+    )
+    def generate(self, image_b64: str, prompt: str) -> Union[str,str]:
+        # fetch model client
+        client = self._get_model_client()
+
+        # create payload
+        payload = self._make_payload(prompt, image_b64)
+
+        # generate response
+        start_time = time.time()
+        answer, finish_reason, usage = self._generate_response(client, payload)
+        end_time = time.time()
+
+        # collect statistics on prompt, usage, and timing
+        stats = {
+            'api_call_duration': end_time - start_time,
+            'prompt': prompt,
+            'usage': usage,
+            'finish_reason': finish_reason,
+        }
+
+        return answer, stats
```

Comment:
I think this is good. How would we implement the ASKEM equation extractor using this framework? @Tranway1